Generation of constants for peripheral information

FawazTirmizi opened the pull request:
This PR adds support for build-time generation of peripheral constants.

FawazTirmizi committed:
Adds build utils for peripheral information

Adds two functions to build_util: `task_peripherals()` and
`task_peripherals_str()`. The former returns a BTreeMap of the
peripheral names and a struct with their base address and size. The latter
takes that information and generates constants for them of the name
`{peripheral}_BASE_ADDR` and `{peripheral}_SIZE`, returning it as an
`std::String`.

FawazTirmizi committed:
Switched PLIC server to get base addr from xtask

Before, the `build.rs` would need to parse the `chip.toml` in order for
it to get the information it needed. Now, it can use the new functions
in `build_util` to get it.

beezow commented on build/util/src/lib.rs:
```diff
@@ -82,6 +84,44 @@ pub fn task_maybe_config<T: DeserializeOwned>() -> Result<Option<T>> {
 }
 
+pub fn task_peripherals() -> BTreeMap<String, Peripheral> {
+    ron::de::from_str(&env::var("HUBRIS_TASK_PERIPHERALS").unwrap()).unwrap()
```
Can we get a more verbose error message for this.  The unwraps are `env::var` are really hard to debug if they fail.  In my experience it just throws a generic `environment variable not found` and even `RUST_BACKTRACE=1` doesn't help.

FawazTirmizi replied:
Done in next revision

beezow commented on build/xtask/src/config.rs:
```diff
@@ -287,6 +287,28 @@ impl Config {
+                Some(peripheral) => task_peripherals
+                    .insert(name.to_string(), peripheral.clone()),
+                None => panic!(
+                    "Task {} uses a non-existing peripheral",
```
Could we also include the peripheral name that the task was trying to use?

FawazTirmizi replied:
Done in next revision

furquan-rivos commented on drv/riscv-plic-server/build.rs:
```diff
@@ -41,15 +29,6 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
     };
 
-    let plic_base: u32 = {
-        let mut path: String = env::var("HUBRIS_CHIP_DIR").unwrap();
```
Are you planning to revert [this change](https://github.com/rivosinc/hubris/commit/3012e07b98df41ee159f2935d201786cd3dff193) for `HUBRIS_CHIP_DIR` now?

FawazTirmizi replied:
Ah yes, I'll add that on into this patch series.

beezow commented:
Is there an example anywhere of how to use these constants? Perhaps propagate the new api to the FE310 rtc or something?

FawazTirmizi commented:
> Is there an example anywhere of how to use these constants? Perhaps propagate the new api to the FE310 rtc or something?

The FE310 RTC doesn't use the u32 constants, but instead takes the base pointer and uses them to generate the PLIC struct pointer. I could add an example in the comments, if that would suffice.

FawazTirmizi committed:
Revert "Export the path to the chip directory"

This reverts commit 3012e07b98df41ee159f2935d201786cd3dff193.

With peripheral information now being exposed to tasks through
`build-util`, there's no point in exporting this path.

beezow commented:
> > Is there an example anywhere of how to use these constants? Perhaps propagate the new api to the FE310 rtc or something?
> 
> The FE310 RTC doesn't use the u32 constants, but instead takes the base pointer and uses them to generate the PLIC struct pointer. I could add an example in the comments, if that would suffice.

[Here](https://github.com/rivosinc/hubris/blob/1822ed2663d170639a69c51c996dae5e044740f1/drv/fe310-rtc/src/main.rs#L29) It uses a constant from the chip.toml.  Couldn't this use the constants generated by this PR?

FawazTirmizi commented:
> > > Is there an example anywhere of how to use these constants? Perhaps propagate the new api to the FE310 rtc or something?
> > 
> > 
> > The FE310 RTC doesn't use the u32 constants, but instead takes the base pointer and uses them to generate the PLIC struct pointer. I could add an example in the comments, if that would suffice.
> 
> [Here](https://github.com/rivosinc/hubris/blob/1822ed2663d170639a69c51c996dae5e044740f1/drv/fe310-rtc/src/main.rs#L29) It uses a constant from the chip.toml. Couldn't this use the constants generated by this PR?

I got the FE310-RTC and PLIC mixed up. You're right, I'll add it.

FawazTirmizi committed:
Updated FE310-RTC driver to use new periph consts

Now that `build_util` can provide constants for peripherals, there is no
need to hardcode the address in the driver.

furquan-rivos commented on drv/fe310-rtc/build.rs:
```diff
@@ -0,0 +1,15 @@
+use std::env;
+use std::fs::File;
+use std::io::Write;
+use std::path::PathBuf;
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let mut file = {
+        let out = PathBuf::from(env::var_os("OUT_DIR").unwrap());
+        File::create(out.join("rtc_config.rs")).unwrap()
+    };
+
+    writeln!(file, "{}", build_util::task_peripherals_str())?;
+
+    return Ok(());
+}
```
Do you see value in having a helper function that generates the file directly instead of having each component do this? I am thinking we can avoid a lot of boilerplate code then:

```
fn main() -> Result<(), Box<dyn std::error::Error>> {
    build_util::generate_peripheral_consts_file();
}
```

FawazTirmizi replied:
I thought about that as well. I think we'd want it to take in the name as an input, but otherwise something like that is probably a good idea. Maybe out of scope for this PR though? Not a lot of work so either way it's fine with me.

furquan-rivos replied:
> I think we'd want it to take in the name as an input, but otherwise something like that is probably a good idea.

You don't really need that as well. We can standardize on the name, something like "periph_consts.rs". 

> Maybe out of scope for this PR though? Not a lot of work so either way it's fine with me.

I am fine either ways.

FawazTirmizi replied:
I'll save it for a later PR, then.

furquan-rivos replied:
Ack.

furquan-rivos reviewed (dismissed):
Marking approved in case you decide to handle the build_util change as a follow up.

FawazTirmizi commented on drv/fe310-rtc/build.rs:
```diff
@@ -0,0 +1,15 @@
+use std::env;
```
uh oh spaghettios

FawazTirmizi commented:
Just rebased on top of `rivos/main`.
